### ISSUE TITLE
Commands: TRIM, STOP, SEQUIN

### ISSUE THREAD

**Issue by lexelby** [lexelby/inkstitch#24]
It would be very useful to be able to add instructions like TRIM, STOP, or SEQUIN (drop a sequin, supported by certain machines).  Some questions are up in the air, such as:

* does libembroidery/embroidermodder support all of these?
* what should the UI look like?

**Comment by lexelby** [comment]
Looking deeper at libembroidery (which is in embroidermodder), it _does_ have support for TRIM, STOP, and SEQUIN.  Well, sorta.  SEQUIN exists only as a theoretical option but none of the machine format modules actually supports it.  Still, it's a start, and I may be able to add support to libembroidery.  TRIM and STOP support is fully there in libembroidery.  The CSV file uses "COLOR" to indicate "STOP", but they're truly the same thing under the hood.

@wwderw The important question I have for you is: what is your preferred embroidery file format?  You suggested in the other thread that I aim for DST, so that's where I'll start.  Does DST support the "drop sequin" command?

**Comment by wwderw** [comment]
>  The CSV file uses "COLOR" to indicate "STOP", but they're truly the same thing under the hood.

This is true for your home machines that use just one color.  A color change on a single needle would logically dictate a stop.  On multi-needles it isn't quite as straight forward.  It would be a two step process if a color change is used to dictate a stop.  It's possible to do this, but not quite as straight forward.  It would also involve the color assignment step at the machine.

For a lot of machines, TRIM command is actually multiple JUMP commands right in a row.  How I got TRIMs in my other design is create 3 manual jumps, that I went into the CSV file and later changed them to JUMP.

For Sequins, I think there is going to need to be 2 commands, one for Sequin On and then Sequin Drop.  I'm speculating, but I'm thinking that's the case, because the sequin dispenser on my Barudan can be engaged quite a bit sooner then when the first sequin drops.  But I don't have the means of dissecting my files to determine if this is the case or not.  Stop command would also come in handy here, since I have a single dispenser, I have to do a stop command in order to change the thread and/or sequin shape/color while the design is active.

I'm just thinking tick boxes to engage specific functions for certain objects that they would apply to.

Now, depending on what version of software I use, sequin functionality either has to be digitized manually or you can actually use a fill and apply sequins to that fill, using a pre-determined tack down stitch for the sequin and knowing the size of said sequin, overlap etc.  Most sequins used in the US are 5mm.

That would be cool if it could be automated like that, but I'm sure that's a feature for a later time if it's possible.

>The important question I have for you is: what is your preferred embroidery file format? You suggested in the other thread that I aim for DST, so that's where I'll start. Does DST support the "drop sequin" command?

Yes, DST can handle that specific command.  I've used DST files for that as well.  The only real function that I personally would like, but is very, very machine specific (or a function in proprietary software) is slow/fast commands.  In my case that would be the Barudan U?? file.

I would just get the DST format loaded for bare is that is the most universal file format that we have (even with quite a lot of the home machines as well, although sequin support would be more of a proprietary nature done through their software, but the others should be able to be handled even at the home machine).

**Comment by lexelby** [comment]
> This is true for your home machines that use just one color. A color change on a single needle would logically dictate a stop. On multi-needles it isn't quite as straight forward. It would be a two step process if a color change is used to dictate a stop. It's possible to do this, but not quite as straight forward. It would also involve the color assignment step at the machine.

Hmm, okay.  libembroidery's inernal representation has a "STOP" command, and the CSV output writes this as "COLOR".  But now that I look at it again, I think it's just as you say, that this _does_ indicate a color change.

> But I don't have the means of dissecting my files

I do!  Great idea.  When you have some free time, can you create me a few DST files, showcasing the various commands we're looking for here?  Something like:

* one file with some running stitch, a STOP, and then more running stitch
* one file with some running stitch, a TRIM, and then more running stitch
* one file demonstrating a SEQUIN
* one file demonstrating the sequin fill you described

In theory I should be able to reverse engineer those and add the necessary features to the DST outputter in libembroidery.

I'm also considering doing away with ink/stitch's output module entirely in favor of having it interface with libembroidery directly.  That way ink/stitch could directly output any format that libembroidery supports!  This should be fairly easy to do, and I can just package libembroidery into the pyinstall-ified package like any other library.

**Comment by wwderw** [comment]
Attached are all the functions in separate files that I was talking about.  4 DSTs and 1 U?? file.

[Stitch Files.zip](https://github.com/lexelby/inkstitch/files/1629030/Stitch.Files.zip)

>I'm also considering doing away with ink/stitch's output module entirely in favor of having it interface with libembroidery directly. That way ink/stitch could directly output any format that libembroidery supports! 

That's a cool idea.  I don't know how easy it would be able to do, but might want to think about a batch exporting function as well.  When I do my stock designs, I usually have to save in Wilcom one at a time in the individual format.

I don't know how possible that is, just got me thinking when you mention the tighter integration with the conversion module from EmbroiderModder.

**Comment by lexelby** [comment]
@wwderw  Shoot, I forgot.  For comparison, can you also send a DST that shows a color transition without a stop?

**Comment by wwderw** [comment]
I'm thinking that they do something with the stop command in the multiple jump sections at the end of the object.  But that is me speculating.  I would be curious to know what they do indeed do different.

[Color Change Only.zip](https://github.com/lexelby/inkstitch/files/1629817/Color.Change.Only.zip)

**Comment by lexelby** [comment]
Wow, DST sure is confusing.  I wish it wasn't the de facto standard format.

After digging around in a hex editor, I have so many questions for you:

1. What embroidery software did you use to create those DST files?
2. Three jump stitches make a TRIM in DST.  Does it matter what the coordinates of those jump stitches are?
3. Do you know why there are so many jump stitches at the end of the "Color Change Only" file?
4. Upon examining them, the "Color Change Only" and "Run With Trim" look effectively identical.  What exactly do they do differently when you sew them out with your machine?
  * I can't see anything that differentiates the "stop" in one from the "color change" in the other -- they have the same bits set.  My only theory is that maybe they've somehow encoded it in the positions of the stitches before or after the stop command?
5. Sequins in the DST are even more confusing.  I can't for the life of me figure out what in the DST tells the machine to load or drop a sequin.  Is there possibly a secondary file with more data...?
6. Maybe there's another format we should be focusing on instead...?

All that said, I've already ripped out the CSV-writing part of inkstitch and replaced it with libembroidery.  It works really well!  I'm in the process of adding "jump after", "trim after" and "stop after" checkboxes to the Params UI.

Those checkboxes will just make use of libembroidery's JUMP/TRIM/STOP settings.  Our results will entirely depend on how well libembroidery's code supports those features for a given embroidery file format.

For DST, I can tell just by looking at the code that it doesn't support TRIM and it does nothing to differentiate between a STOP and a color change.  I can easily modify the code and contribute a fix to embroidermodder, but only once we've gotten to the bottom of this puzzle about the file format.

**Comment by lexelby** [comment]
More questions:

7. I noticed that your files have little back-then-forward bits at the end of each run of stitches.  I'm assuming that is to tie off the thread.  Did you insert that purposefully yourself?  Did your software do it for you?  Or maybe I'm wrong and those backtracks are part of some secret data encoding and they're not actually treated as stitches?

**Comment by wwderw** [comment]
>What embroidery software did you use to create those DST files?

Wilcom

>Three jump stitches make a TRIM in DST. Does it matter what the coordinates of those jump stitches are?

Ideally you want them heading in the direction of the next stitching object, but otherwise, I do not believe so, just so that you have 3 coordinates to have jump functions to invoke a trim

>Do you know why there are so many jump stitches at the end of the "Color Change Only" file?

Some I believe are what Wilcom calls "empty" jumps and are redundant.  Why they are there, I couldn't tell you.

>    Upon examining them, the "Color Change Only" and "Run With Trim" look effectively identical. What exactly do they do differently when you sew them out with your machine?
    I can't see anything that differentiates the "stop" in one from the "color change" in the other -- they have the same bits set. My only theory is that maybe they've somehow encoded it in the positions of the stitches before or after the stop command?

I would have to agree with that theory, because on the printouts of files with stops, it will show black, black with stop (usually the single stitch that the stop is at) and then black (or whatever color is next in line).

At my machine, doing the color assigning part, it will have one for the color, the stop will already be using the machines stop function and then you'll have your next color if there is a color change or it will continue on.  So it must be hard coded in there on that last stitch.  I'll see if I can get a PDF for you and that should tell you the exact stitch that it affects.

>Sequins in the DST are even more confusing. I can't for the life of me figure out what in the DST tells the machine to load or drop a sequin. Is there possibly a secondary file with more data...?

It exports one file, I could take a screen shot of the Stitch list that Wilcom shows (akin to CSV file, but not a dedicated CSV file that's exported or anything like that.

>For DST, I can tell just by looking at the code that it doesn't support TRIM and it does nothing to differentiate between a STOP and a color change. I can easily modify the code and contribute a fix to embroidermodder, but only once we've gotten to the bottom of this puzzle about the file format.

3 JUMP functions in a row will invoke a trim.  So if someone ticks the trim box, just have the program now to generate 3 jump stitches in the direction of the next object to be digitizing.  Doesn't really matter how far out etc, I would just make sure that it's in the direction of the next object.

>I noticed that your files have little back-then-forward bits at the end of each run of stitches. I'm assuming that is to tie off the thread. Did you insert that purposefully yourself? Did your software do it for you? Or maybe I'm wrong and those backtracks are part of some secret data encoding and they're not actually treated as stitches?

Those are tie off stitches.  Those are going to be needed before a trim is done.  Most commercial digitizing software will do that on their own.  If I'm using the manual stitch (one mouse click equals 1 needle penetration), I would have to do those (tie in (at the start of an object) and tie off (at the end of an object) on my own.  

We would want both sets in there.

**Comment by lexelby** [comment]
Thanks for the details!  I'll get the TRIM functionality added in.  It's going to make the most sense to fix that in libembroidery, since we don't want to add 3 trims for _other_ formats that have proper trim commands built in.

The PDF and the stitch list that you mentioned would both be super helpful, thank you!

**Comment by wwderw** [comment]
See if that helps.

[Files.zip](https://github.com/lexelby/inkstitch/files/1633434/Files.zip)

**Comment by wwderw** [comment]
I'm just mentioning this while I'm thinking about it.  I've gotten to the point that if I don't mention things while I'm thinking of them, chances are they will be gone forever.

I remember you saying this:

>Sequins in the DST are even more confusing. I can't for the life of me figure out what in the DST tells the machine to load or drop a sequin. Is there possibly a secondary file with more data...?

The trials and tribulation that we were having with the stop code got me thinking about how the machine does the sequin function.  I got to thinking if it's somewhat along the same lines as the stop with color changes.  So I did a 5 sequin run.

I've attached the DST file, the CSV version of the DST file, the DGT file (which is just a bitmap of the thread design), picture of the stitch list from my program and the picture of how it looks like on the control panel of my Barudan machine.  This picture is of where I do the needle assignment.  C01 means needle 1 (what the sequin dispenser uses), S1 and S0 are Sequin On and Sequin Off respectively.

Based on the CSV version, C01, S1 and S0 are done with Color functions and jump codes are used for one it actually dispenses the sequin.  Now, that's based on how libembroidery interpreted everything.  

Now S1 and S0 were already determined by the file itself, it had already told the machine to do those functions on this list.  I can easily get to that on my own if need be.  Since I don't send out sequin embroidery files, that's not an issue.

The pattern of the stitch work before and after the sequin is dropped would be something that I would suggest be done in the extension.  It's a basic pattern, but it would help if the extension/inkscape did that heavy lifting itself so it was manual.  Assume 5mm sequins as standard, use the same measurements for the stitches.

As far as getting the extension to know when to do sequins, I was thinking about maybe a line with a stroke **combined** with a circle (5mm) that only has a stroke.  The extension would know that that object is for sequin functionality.

Thoughts?  Does this help at all?

[Stitch Files.zip](https://github.com/lexelby/inkstitch/files/1650390/Stitch.Files.zip)

**Comment by lexelby** [comment]
Yes, this helps a _ton_.  I'm going to save the sequin feature for later since it's clearly going to be complex.  How often do you need sequin functionality?  By which I mean, how highly do you think I should prioritize the sequin feature?

Say, does your machine support those awesome flippable sequins?  The ones where it has different colors on each side and the hole is near the side.  My kids love those :)

**Comment by wwderw** [comment]
> By which I mean, how highly do you think I should prioritize the sequin feature?

At best, tertiary importance.  I do use it for embellishments on patterns, but not as often as I would need other functions such as stop, trim, batch export, pdf printout w/stitch list etc.

>Say, does your machine support those awesome flippable sequins? The ones where it has different colors on each side and the hole is near the side. My kids love those :)

Unfortunately no.  Only center mass holes are what my machine can take.  I can't move the needle like you can on a sewing machine from side to side.

![hat sequin](https://user-images.githubusercontent.com/24878362/35203772-8265becc-feef-11e7-9744-0061577a73f7.jpg)

**Comment by lexelby** [comment]
Okay, I had a chance to examine the files, and it's exactly as you say.  When you tell the machine S1, after that it drops a sequin for every JUMP.  Easy peasy!

Here's the only bit that trips me up:

> Now S1 and S0 were already determined by the file itself, it had already told the machine to do those functions on this list.

I really can't see how the DST could tell the machine about the S1 and S0 at all.  There's nowhere in the DST to specify colors at all.  I checked just to be sure and didn't find anything extra at all.

How do you get files from Wilcom into the machine?  Are you putting them on a flash drive (or floppy disk...?) and carrying them over?  Or are they transferred by USB?

I'm wondering if Wilcom is sending over color information separately somehow.  There are a couple of common embroidery color list formats like .COL and .INF that it might be sending, or it might use some special communication protocol to tell the machine directly.  What do you think?

**Comment by wwderw** [comment]
I move the DST file on a CF card to the machine.  The only other file that is rendered at the time of exporting is a DGT file, which is nothing more then a bitmap file and typically I have that option disabled.  I only included this time due to trying to figure out how to implement this in the extension.

As far as how the file is able to tell the machine to already have S1 and S0 at the correct locations, that I couldn't tell you.

Honestly, myself, I wouldn't be worried about it.  As long as the information is relayed to the operator (ie PDF sheet) that mentions this info, it is not any more or less difficult then assigning needles (or stops) at the machine.  It's all done at the very same menu.  As long as the operator knows what S1 and S0 mean, all is good.

Since I usually don't send out DST files with sequins (as none of my customers to my knowledge have sequin dispensers), I personally don't have a problem with having to manually assign, just so long as there is something to help remind me where it is in the stitch order.

**Comment by lexelby** [comment]
Alright, we can leave it for now. :)  Ultimately I'd like to try to get to the bottom of this in case it might impact others' workflows more significantly than yours, but it's a fairly low priority.

**Comment by tatarize** [comment]
>>does libembroidery/embroidermodder support all of these?

No. Josh didn't really attach or decode any sequin stuff anywhere. He also doesn't differentiate between COLOR_CHANGE and STOP. 

>>I can't see anything that differentiates the "stop" in one from the "color change" in the other -- they have the same bits set. 

This is false. The 3rd byte in the stop sequence is b[2]  has everybit 0b11110011 set. The 5,6 bit in the color change sequence of b[2] has everybit 0b11000011 set.

0b11110011
0b11000011

So if bits 5,4 are set, it's a stop and but not a color change. 
0b10000011 is a move. 

And the dx dy stuff reads bits 2,3 and 5,4 respectively for the +81, -81 value in those coordinates.

So a stop without a color change is a stop flagged with +81 y and -81 y. 

The reason for it to seem confusing is there is this format comes from punch cards. And is setup so that they are easy to punch, and if you punch the wrong thing, you can correct it. So rather than 2s complement it has a weird greedy algorithm setup where you define things in sets of + and - values in 3n of the previous value. 1, 3, 9, 27, 81. So there's a lot of different ways that define 0. 

The command for engaging the sequin is:
0b01000011

You can see the full listing (as best I can tell in my decoding fiddle): https://jsfiddle.net/sL8z1vmh/


----

The color information for needle sequence is usually entered into the embroidery machine. Usually you give the sequence of the needles and it will engage them in the order you give. The file itself only tells whether to move on to the next needle in sequence or not (and even that is just a hack into the format itself)

**Comment by lexelby** [comment]
Hi @tatarize, welcome to the project!

Interesting.  I thought 0xF3 (0b11110011) was the "end of embroidery" sequence.  This page and embroidermodder's `format-dst.c` seemed to bear that out: http://www.achatina.de/sewing/main/TECHNICL.HTM.  All of the example files that @wwderw gave me (generated by wilcom) reflected that.  When he adds a "STOP" code to his embroidery design and saves it as DST with wilcom, it doesn't have a 0b11110011 at that spot, just a 0b11000011.

After a lot of back-and-forth, we eventually realized that his machine doesn't automatically understand a STOP at the spots where he enters stop codes.  Rather, he sets the needle for that block to `C00` (normal needles are `C01` - `C15`).  `C00` (on a Barudan machine) says "sew with the last color and pause at the next color change".

Thanks for the tip on sequin enable/disable being `0b01000011`!  That matches the example files that @wwderw gave me.

Someday maybe I'll get a "real" machine so I can experiment with all of this :)

**Comment by tatarize** [comment]
0xF3 is stop.  It's also typically end. Really the difference is at the end, your machine stops without a color change. Wilcom software will rather than print out three bytes will print out 1 byte of 0x1A at the end, after the 0x0000F3, but I've never seen any software actually care. It's again, designed for single-needle punchcard reading machines. So it keeps reading data while there is data and the machine does what the command is, and just what the command is. You do very much put a 0xF3 at the end of the embroidery but it's not actually telling the machine to end it tells it to stop. 

In single needle machines stop and color_change and end are also the same command. Namely because they don't do anything, other than stop the machine and wait for the user to start it back up again. In the case of color_change it's because they are switching out the thread. In the case of STOP they could be messing with applique or some such thing. In the case of END, they won't be starting the machine back up, but all of these commands are the same thing and they all actually read the same. The third bit tests positive for the bits 11xxxx11. My guess is when the multineedle machines came around they started using stop as color-change until they started needing something for general stop so they set the +81 and -81 in the y coord to indicate the distinction. Though, oddly enough I'd need access to some dst files on punch cards to test that theory. Though I seriously doubt 11110011 predates 11000011.

The harder concern is how to mark up the stitches themselves with particular commands in SVG. When I had considered using SVG as a baseline format for Touch Embroidery and explored it that was one thing I couldn't ever do.  It was one of the major limitations among others that had me abandon the idea (speed being another one). The other major issue was I wanted a sort of universal format that open/closed sourced projects could all utilized equally and there's a pretty hefty set of unintended but overt limitations to the "vector-object + fill = embroidery" paradigm. 

Going forward you'll also find that libembroidery has a number of oversights and flaws. Even something as simple as taking a .DST and converting it into a bunch of fairly common formats and loading those with something like Wilcom TruSizer (free Wilcom conversion software) will find some of them are still quite rough. While fantastic work, I had to rewrite them for some good consistency and even then some of the formats don't actually work for my users. Add to this the rather interesting fact that @wwderw would be happy to verify basically every single embroidery machine in existence will always natively read .dst files even if they are not tajima machines themselves. If you get the manual to any machine and check, they (as far as I know) always support .dst files natively.

This is, in part, why some of the commands seem hacky and weird. Like multiple jumps equating to a trim. So many creating formats will take trim and convert it to +2 +2, -4, -4, +2 +2 to ensure the trim, then do a bunch of jumps to end up where it would have cut anyway. There was no trim command initially. This command was a later hack into the format. It's like the worst consensus format ever.

There may even be some machines with a bit of other hacky stuff added that didn't really get popular. So the machine itself could hack in the encoding for various other commands and have never become popular enough that people started writing that. I daresay 0b11111111 as a command would be treated as a stop and 0b11001111 would be treated as a colorchange, but you could certainly interpret them differently. You have some 24 bits that encode position and so 12 different bit sets that cancel out to zero. So some 4096 ways to trigger "go nowhere", 1 of which when the control codes are engaged differences between stop and color_change.

**Comment by wwderw** [comment]
> Add to this the rather interesting fact that @wwderw would be happy to verify basically every single embroidery machine in existence will always natively read .dst files even if they are not tajima machines themselves. If you get the manual to any machine and check, they (as far as I know) always support .dst files natively.

Your mid-range **home** machines may or may not read DST files natively, even today.  Your low end **home** machines are almost exclusively the native format of the OEM vendor.

For high end home/entry level commercial to full fledge commercial, DSTs are the standard.

I've known some PR owners (high end home/entry level commercial Brother machine) that didn't know that they could read any format other then PES, but I would imagine that's due to the place that they got training on the machine.

**Comment by tatarize** [comment]
Wilcom dst files seem to end. 00 00 F3 1A. I doubt there's any real distinction between 00 00 F3 and 00 00 C3 there. Both mean stop. Though the latter means go on to the next needle in sequence too, which wouldn't exist and thus do nothing, or does and then would still basically do nothing. 

You'll still get plenty of mileage out of SVG. There's plenty of stuff that easily fits that paradigm and should work quite well. But, a there's other stuff that won't.  For example, guilloche are often Eulerian and thus entirely doable in embroidery. Or things like L-System fractals that are equally able to be done by these same methods. You can simply stitch them out and I can and have done this. However, you cannot describe those in terms of fills attached to shapes. 

![0ceb5045e8d5b0b98d6c6ecccaa39d6c](https://user-images.githubusercontent.com/3302478/41450381-1a70ae3c-701c-11e8-931c-aa0ab6e3e4ac.jpg)

You can do quite a lot, but you miss some stuff and it's not anything to do with SVGs mind you. In fact, implement a solid and robustly defined namespace and I'd likely be happy to add it to some programs I oversee. Rather the problem seems to be with the paradigm itself. The idea that embroidery is well defined by adding fills to vector shapes. 

I modified the transcoded to Java libembroidery files that got added to Mobile Viewer, to fix these issues and released them. And opened issues on Embroidermodder ( https://github.com/Embroidermodder/Embroidermodder/issues/91 https://github.com/Embroidermodder/Embroidermodder/issues/109 ) as well as did complete tear downs of .PES format. I think JEF has an error that stops neither export nor import of the file as where it writes  0x0A seems like it should write 0x14. A user reported it doesn't load on a machine that took .JEF format files, but I have a 15 needle SWF machine that doesn't take .JEF, so I can't really test such things. Without testing the files on physical machines the best you can do is make them look like the files Wilcom produces but even those, apparently, have some errors (I cross checked PES files with Brother PE software and found some issues in the defined palette methodology used).

I've not seen any modifications to .dst beyond the basic ones to add in a stop without color change and to add in sequin commands.

**Comment by tatarize** [comment]
Using just SVG mind you, you could duplicate most of the plausible work done by Brother PE-Design software line. The breakdown of the format (note this part was Josh and not me) shows that it's just taking geometric objects and tagging them with a bunch of properties and using the engine to re-create the stitches. https://github.com/frno7/libpes/wiki/PES-object-section#block-description

Also, the speed issue was largely due to the way files were loaded and saved and I needed it to go so fast that it could load and save the entire embroidery so rapidly as to implement a undo/redo system for my app, as the previous methods of declaring each operation performed and saving the data needed to reverse that operation was too unwieldy. 

My objection is largely to the paradigm itself not to your perfectly fine and much needed implementation of it. In fact, if you could define a namespace with all the sorts of fills etc, that would be quite useful even if the underlying engine for how the stitches are laid down from those specifications were largely less defined. I very seriously considered doing that myself until the various counter-examples of valid embroidery that couldn't be captured by the scheme got to me, or things like how embroidery designers coax the software to do what they want and get the tatsumi elements in the right order or to get an underlay of a cluster of objects and then fill each object in sequence would require creating a vector shape that matches the entire shape to do that underlay as a very sparse fill. Or the sheer amount of talent and expertise it requires to do blending at the edges between different objects by making them bigger and adding random edges. and blending the fill pattern, etc.  When in reality these are things that should be handled for you, but are simple consequences of the paradigm of needing a shape to do a fill.

That said, you should overtly define the start and end locations of a flat fill with a start and end location rather than trying to coax some already existing svg elements into performing that role. Shapes have start points and end points and they shouldn't be jiggered so you avoid saving a tiny amount of data. 

I addressed some of these points a couple years ago, on the Embroidermodder issue asking about an extension of svg as such: https://github.com/Embroidermodder/Embroidermodder/issues/65#issuecomment-204657094

Don't take my somewhat downtrodden views on the design path as anything personal. It's completely good and solid work. You could clearly make a program at least as good as those programs out there that sell for the price of a car, using exactly this scheme (in fact your way further along than most such attempts). It's entirely possible and reasonable to define a totally good and working namespace that could define extended elements to the SVG format such that it could capture the vast majority of modern embroidery design in both an open and important way. You could easily define a sort of standard extension to SVG to define most of the commonly used fills, and create a workable standard for other projects and an interface language for PESv6 <-> SVG as some folks have wanted to do. I just think that "vector shape + fill = embroidery" is a sometimes useful shorthand to define some embroidery but far from a meaningful description of all possible embroidery. But, fantastic work.

**Comment by lexelby** [comment]
I think that perhaps it's more possible than you realize to implement your preferred paradigms of freehand drawing and mathematical functions (l-systems, guilloche, etc) in Ink/Stitch.  We'll have to see what we come up with. :)  Feel free to submit PRs if you find yourself motivated to do so.

It's interesting that you translated libembroidery to Java.  I've been toying around with the idea of translating it to Python, to make modifications easier and to solve an issue we're having with packaging the libembroidery binary for MacOS.  Did you run into much difficulty with the translation work?
